### PR TITLE
[smartmeter] Fix trying to create channels from invalid data

### LINKED ISSUE #6762: [smartmeter] Binding hangs after protocol error
Hello,
I am using the smartmeter binding version 2.5.0.

## Current Behavior
Sometimes my IR Reader catch up some wrong characters which lets the smartmeter parser crash. After this, this binding does not handle any data. I can restart this binding when I disable the binding in the paperui and enable it afterwards.
My reader produce this error one or two times the day. I am not sure why it happens at the moment, but this is an other issue.
I have configured the binding via rfc2217 and protocol D.

Here the stacktrace.

`java.lang.IllegalArgumentException: ID segment '1-0_56-7-0_2^B^B^Z
#s
^CR)^M
1-0_76-7-0_255' contains invalid characters. Each segment of the ID must match the pattern [A-Za-z0-9_-]*.
        at org.eclipse.smarthome.core.common.AbstractUID.validateSegment(AbstractUID.java:97) ~[bundleFile:?]
        at org.eclipse.smarthome.core.common.AbstractUID.<init>(AbstractUID.java:75) ~[bundleFile:?]
        at org.eclipse.smarthome.core.common.AbstractUID.<init>(AbstractUID.java:58) ~[bundleFile:?]
        at org.eclipse.smarthome.core.thing.UID.<init>(UID.java:57) ~[bundleFile:?]
        at org.eclipse.smarthome.core.thing.type.ChannelTypeUID.<init>(ChannelTypeUID.java:50) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.SmartMeterChannelTypeProvider.getChannelType(SmartMeterChannelTypeProvider.java:85) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.SmartMeterChannelTypeProvider.valueChanged(SmartMeterChannelTypeProvider.java:75) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$9(MeterDevice.java:220) ~[bundleFile:?]
        at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_191]
        at org.openhab.binding.smartmeter.internal.MeterDevice.addObisCache(MeterDevice.java:218) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21MeterReader.populateValueCache(Iec62056_21MeterReader.java:58) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21MeterReader.populateValueCache(Iec62056_21MeterReader.java:1) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$6(MeterDevice.java:184) ~[bundleFile:?]
        at io.reactivex.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:65) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97) [bundleFile:?]
        at io.reactivex.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:100) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenSourceSubscriber.onNext(FlowableRepeatWhen.java:159) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRefCount$RefCountSubscriber.onNext(FlowableRefCount.java:199) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.dispatch(FlowablePublish.java:547) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.onNext(FlowablePublish.java:231) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableTimeoutTimed$TimeoutSubscriber.onNext(FlowableTimeoutTimed.java:101) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableCreate$NoOverflowBaseAsyncEmitter.onNext(FlowableCreate.java:403) [bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21SerialConnector$1.newDataMessage(Iec62056_21SerialConnector.java:103) [bundleFile:?]
        at org.openmuc.j62056.Iec21Port$ModeDReceiver.run(Iec21Port.java:71) [bundleFile:?]
2020-01-04 22:33:03.863 [ERROR] [ding.smartmeter.internal.MeterDevice] - Meter listener failed
java.lang.NullPointerException: null
        at org.openhab.binding.smartmeter.internal.SmartMeterChannelTypeProvider.getChannelTypeIdForObis(SmartMeterChannelTypeProvider.java:113) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.SmartMeterHandler$1.valueChanged(SmartMeterHandler.java:168) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$9(MeterDevice.java:220) ~[bundleFile:?]
        at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_191]
        at org.openhab.binding.smartmeter.internal.MeterDevice.addObisCache(MeterDevice.java:218) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21MeterReader.populateValueCache(Iec62056_21MeterReader.java:58) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21MeterReader.populateValueCache(Iec62056_21MeterReader.java:1) ~[bundleFile:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$6(MeterDevice.java:184) ~[bundleFile:?]
        at io.reactivex.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:65) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97) [bundleFile:?]
        at io.reactivex.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:100) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenSourceSubscriber.onNext(FlowableRepeatWhen.java:159) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRefCount$RefCountSubscriber.onNext(FlowableRefCount.java:199) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.dispatch(FlowablePublish.java:547) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.onNext(FlowablePublish.java:231) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableTimeoutTimed$TimeoutSubscriber.onNext(FlowableTimeoutTimed.java:101) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableCreate$NoOverflowBaseAsyncEmitter.onNext(FlowableCreate.java:403) [bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21SerialConnector$1.newDataMessage(Iec62056_21SerialConnector.java:103) [bundleFile:?]
        at org.openmuc.j62056.Iec21Port$ModeDReceiver.run(Iec21Port.java:71) [bundleFile:?]
2020-01-04 22:33:04.903 [ERROR] [ding.smartmeter.internal.MeterDevice] - Fatal error occured
io.reactivex.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | java.lang.IllegalArgumentException: UID segment '1-0_56-7-0_2^B^B^Z
#s
^CR)^M
1-0_76-7-0_255' contains invalid characters. The last segment of the channel UID must match the pattern '[\w-]*|[\w-]*#[\w-]*'.
        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704) [bundleFile:?]
        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701) [bundleFile:?]
        at io.reactivex.internal.subscribers.LambdaSubscriber.onError(LambdaSubscriber.java:79) [bundleFile:?]
        at io.reactivex.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:69) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97) [bundleFile:?]
        at io.reactivex.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:100) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRepeatWhen$WhenSourceSubscriber.onNext(FlowableRepeatWhen.java:159) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableRefCount$RefCountSubscriber.onNext(FlowableRefCount.java:199) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.dispatch(FlowablePublish.java:547) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowablePublish$PublishSubscriber.onNext(FlowablePublish.java:231) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:92) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableDoOnLifecycle$SubscriptionLambdaSubscriber.onNext(FlowableDoOnLifecycle.java:79) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableTimeoutTimed$TimeoutSubscriber.onNext(FlowableTimeoutTimed.java:101) [bundleFile:?]
        at io.reactivex.internal.operators.flowable.FlowableCreate$NoOverflowBaseAsyncEmitter.onNext(FlowableCreate.java:403) [bundleFile:?]
        at org.openhab.binding.smartmeter.internal.iec62056.Iec62056_21SerialConnector$1.newDataMessage(Iec62056_21SerialConnector.java:103) [bundleFile:?]
        at org.openmuc.j62056.Iec21Port$ModeDReceiver.run(Iec21Port.java:71) [bundleFile:?]
Caused by: java.lang.IllegalArgumentException: UID segment '1-0_56-7-0_2^B^B^Z
#s
^CR)^M
1-0_76-7-0_255' contains invalid characters. The last segment of the channel UID must match the pattern '[\w-]*|[\w-]*#[\w-]*'.
        at org.eclipse.smarthome.core.thing.ChannelUID.validateSegment(ChannelUID.java:179) ~[?:?]
        at org.eclipse.smarthome.core.common.AbstractUID.<init>(AbstractUID.java:75) ~[?:?]
        at org.eclipse.smarthome.core.thing.UID.<init>(UID.java:66) ~[?:?]
        at org.eclipse.smarthome.core.thing.ChannelUID.<init>(ChannelUID.java:59) ~[?:?]
        at org.openhab.binding.smartmeter.internal.SmartMeterHandler$1.valueRemoved(SmartMeterHandler.java:234) ~[?:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$10(MeterDevice.java:260) ~[?:?]
        at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_191]
        at org.openhab.binding.smartmeter.internal.MeterDevice.notifyValuesRemoved(MeterDevice.java:260) ~[?:?]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$8(MeterDevice.java:189) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_191]
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:1.8.0_191]
        at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1625) ~[?:1.8.0_191]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_191]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_191]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:1.8.0_191]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:1.8.0_191]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_191]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_191]
        at org.openhab.binding.smartmeter.internal.MeterDevice.lambda$6(MeterDevice.java:189) ~[?:?]
        at io.reactivex.internal.subscribers.LambdaSubscriber.onNext(LambdaSubscriber.java:65) ~[?:?]
        ... 14 more
`
## Expected Behavior
I think the binding should better handle these protocol errors. 
In my opinion this binding should skip all wrong formatted data block and should not hang up.

## Steps to Reproduce (for Bugs)
Use this file and send this content via telnet server to the binding.
[SampleData.txt](https://github.com/openhab/openhab-addons/files/4022557/SampleData.txt)

## Your Environment
Openhab 2.5.0
Smartmerter Binding 2.5.0
Configured a rfc2217 URL and the protocol D

### PR DESCRIPTION
**Pull request by t2000**
Fixes #6762

Signed-off-by: Stefan Triller <github@stefantriller.de>

### CODE REVIEW

**Review by msteigenberger — APPROVED** [review]

**Review by Hilbrand — APPROVED** [review]
LGTM

**Review by Hilbrand — CHANGES_REQUESTED** [review]
3 unit tests failed. Related to your changed?

**Comment by t2000** [comment]
Yes they were indeed related to my changes :(

Good that we had them! @msteigenberger could you please have another look again, since I had to split the method into 2 now because to me it looks like they were used for 2 different purposes. Thanks.

**Review by msteigenberger — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/SmartMeterBindingConstants.java [hunk 1]
```diff
@@ -36,14 +36,18 @@
...
     public static final String OBIS_FORMAT = OBIS_FORMAT_MINIMAL + "*%d";
     public static final String CHANNEL_TYPE_METERREADER_OBIS = "channel-type:" + BINDING_ID + ":obis";
 
-    public static String getObisChannelId(String obis) {
+    public static String getObisChannelIdPattern(String obis) {
         return obis.replaceAll("\\.", "-").replaceAll(":|\\*", "_");
     }
+
+    public static String getObisChannelId(String obis) {
+        return obis.replaceAll("\\.", "-").replaceAll(":|\\*", "_").replaceAll("[^\\w-]", "");
```
[the remark this answers is not in the document]
Comment (t2000) [reply bundles/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/SmartMeterBindingConstants.java]: Indeed, makes sense. I have added another commit with the change. Thanks.

**Review by msteigenberger — APPROVED** [review]

**Comment by msteigenberger** [comment]
Lgtm. Thanks!

**Comment by t2000** [comment]
@Hilbrand Since the tests are green now and the changes have been approved again, may I kindly ask you to click the merge button? Thank you!

**Review by Hilbrand — APPROVED** [review]
LGTM. Good to give me a trigger. Due to the amount of messages I sometimes miss things :wink:

**Pull request merged**